app-backend/db/src/main/scala/TaskDao.scala

```diff
@@ -0,0 +1,208 @@
+import doobie._
+import doobie.implicits._
+import doobie.postgres.implicits._
+
+import java.time.Instant
+import java.util.UUID
+
+object TaskDao extends Dao[Task] {
+
+  val tableName = "tasks"
+  val joinTableF =
+    Fragment.const("tasks join task_actions on tasks.id = task_actions.task_id")
+
+  val cols =
+    fr"""
+     SELECT
+      id,
+      created_at,
+      created_by,
+      modified_at,
+      modified_by,
+      project_id,
+      project_layer_id,
+      status,
+      locked_by,
+      locked_on,
+      geometry
+    FROM
+    """
+
+  val selectF: Fragment =
+    cols ++ tableF
+
+  val listF: Fragment =
+    cols ++ joinTableF
+
+  val insertF: Fragment =
+    fr"INSERT INTO " ++ tableF ++ fr"""(
+          id,
+          created_at,
+          created_by,
+          modified_at,
+          modified_by,
+          project_id,
+          project_layer_id,
+          status,
+          locked_by,
+          locked_on,
+          geometry
+     )
+     """
+
+  def updateF(taskId: UUID, update: Task.TaskFeatureCreate): Fragment =
+    fr"UPDATE " ++ tableF ++ fr"""SET
+      project_id = ${update.properties.projectId},
+      project_layer_id = ${update.properties.projectLayerId},
+      status = ${update.properties.status},
+      geometry = ${update.geometry}
+    WHERE
+      id = $taskId
+    """;
+
+  def appendAction(
+      taskId: UUID,
+      initialStatus: TaskStatus,
+      newStatus: TaskStatus,
+      userId: String
+  ): ConnectionIO[Int] =
+    if (initialStatus != newStatus) {
+      fr"""INSERT INTO task_actions (task_id, user_id, timestamp, from_status, to_status) VALUES (
+          $taskId, $userId, now(), $initialStatus, $newStatus
+          )""".update.run
+    } else {
+      0.pure[ConnectionIO]
+    }
+
+  def getTaskById(taskId: UUID): ConnectionIO[Option[Task]] =
+    query.filter(taskId).selectOption
+
+  def unsafeGetTaskById(taskId: UUID): ConnectionIO[Task] =
+    query.filter(taskId).select
+
+  def getTaskActions(taskId: UUID): ConnectionIO[List[TaskActionStamp]] = {
+    Dao
+      .QueryBuilder[TaskActionStamp](
+        fr"select task_id, user_id, timestamp, from_status, to_status FROM task_actions",
+        fr"task_actions",
+        Nil
+      )
+      .filter(fr"task_id = $taskId")
+      .list
+  }
+
+  def getTaskWithActions(taskId: UUID): ConnectionIO[Option[Task.TaskFeature]] =
+    (for {
+      task <- OptionT(getTaskById(taskId))
+      actions <- OptionT.liftF(getTaskActions(task.id))
+    } yield { task.toGeoJSONFeature(actions) }).value
+
+  def unsafeGetTaskWithActions(taskId: UUID): ConnectionIO[Task.TaskFeature] =
+    for {
+      task <- unsafeGetTaskById(taskId)
+      actions <- getTaskActions(task.id)
+    } yield { task.toGeoJSONFeature(actions) }
+
+  def updateTask(
+      taskId: UUID,
+      updateTask: Task.TaskFeatureCreate,
+      user: User
+  ): ConnectionIO[Option[Task.TaskFeature]] =
+    (for {
+      initial <- OptionT(getTaskById(taskId))
```

Comment:
So I think when we do an update on a task, we need to check if this task is locked or not. To me, this would be better to happen on the endpoint level since such logic does not have much to do with updating the record in DB itself, and if locked, I think it is easier to handle request rejection on endpoint level too. So this comment serves as a reminder in case we forget in the next endpoint card.

Reply:
Also -- locked tasks can be updated, since how else would they become unlocked? I agree there's some work to figure out here in the api

Reply:
@jisantuc good point. I agree.